Digitizer switch usages support

avg-I opened the pull request:
The code could probably be better, but at least it works for me.

avg-I committed:
add support for surface and buttons switches

The switches are described by the corresponding usages in the Digitizer
page.  They are controlled by feature reports.

wulf7 commented on hconf.c:
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
+	u_int value;
+	int error;
+
+	error = hconf_get_switch(sc, arg2, &value);
```
Microsoft does not require configuration page feature reports to be readable:
https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/touchpad-required-hid-top-level-collections
This means that old values should be taken from soft context rather then from hardware. hconf_get_switch() can be used at device_attach() handler to initialize variables.

wulf7 commented on hconf.c:
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
+		return (ENXIO);
+
+	fbuf = malloc(sc->switch_rlen[swtype], M_TEMP, M_WAITOK | M_ZERO);
+	sx_xlock(&sc->lock);
```
hid requests are serialized by transport backends now, so no need to protect them with extra locks.  hconf_set_input_mode_impl() takes sx_lock to serialize soft context updates.

wulf7 commented on hconf.c:
```diff
@@ -149,6 +156,77 @@ hconf_input_mode_handler(SYSCTL_HANDLER_ARGS)
+}
+
+static int
+hconf_set_switch(struct hconf_softc *sc, int swtype, u_int mask)
```
hconf_set_switch() should be called from device_resume handler too as suspend/resume cycle often resets touchpad settings.

wulf7 reviewed (commented):
Looks mostly good. I left minor notes pinned to code lines.

avg-I commented:
Thank you very much for the suggestions!
Time permitting, I will implement all of them and submit a pull request.

I am also thinking about how to integrate this feature into desktop environments.
I took KDE / plasma5 as an example. There, a touchpad can be turned off either via a global shortcut or via an applet. It seems that when it's done KDE sets a property "atom" of an input driver and that's up to the driver what to do next. In the case xf86-input-synaptics it seems that the driver simply starts ignoring input events without actually doing anything with the kernel driver or the hardware. So, I think that it would be nice to plug the switch feature into that. But doing it via an OS-specific sysctl does not seem to be nice. I wonder if there is a more generic interface (like a set of ioctls) for setting HID device features. I tried looking in Linux code but could not find anything of interest.

wulf7 commented:
> I will implement all of them and submit a pull request.

Great! Thank you!

> I wonder if there is a more generic interface (like a set of ioctls) for setting HID device features.

I do not know.
Hidraw/uhid can be used but they are too low level.
Unimplemented evdev's EVIOCSMASK ioctl looks promising, but it is per-client, so it requires support from all intermediate libraries like libevdev and libinput to work.
We can invent new input event codes also, but it requires tight coordination with upstream.

avg-I committed:
Merge remote-tracking branch 'upstream/master' into digitizer-switch-usages-support

avg-I committed:
hconf: unify handling of all touchpad controls provided via feature reports

Those include:
- input mode
- surface switch
- button switch

avg-I committed:
hconf: fix handling of multiple controls sharing the same write-only report

While here, complete renaming of "feature toggle" to "feature control"
completely removing the word toggle as well as abbreviation referring
to it (e.g., "ft").

wulf7 reviewed (approved):
Looks good. Thank you.
